classad/_expression.py

```diff
@@ -0,0 +1,299 @@
+import operator
+from collections import MutableMapping, OrderedDict
+
+import pyparsing as pp
+from typing import Any, Iterable, List, Iterator
+
+from classad._operator import eq_operator, ne_operator, not_operator
+from classad._primitives import Error, Undefined, HTCBool
+from ._base_expression import Expression
+from . import _functions
+
+
+def scope_up(key: List[str]):
+    return key[:-1]
+
+
+class ClassAd(Expression, MutableMapping):
+    __slots__ = "_data"
+
+    def __add__(self, other):
+        return Error()
+
+    __sub__ = __rsub__ = __radd__ = __add__
+
+    def __init__(self):
+        super().__init__()
+        self._data = OrderedDict()
+
+    def __setitem__(self, key: str, value: Expression) -> None:
+        """
+        Keynames that are reserved and, therefore, cannot be used: error, false, is,
+            isnt, parent, true, undefined
+        """
+        try:
+            key = key.casefold()
+        except AttributeError:
+            key = key._expression.casefold()
+        if key in ["error", "false", "is", "isnt", "parent", "true", "undefined"]:
+            raise ValueError(f"{key} is a reserved name")
+        self._data[key] = value
+
+    def __delitem__(self, key: str) -> None:
+        self._data.pop(key, None)
+
+    def __getitem__(self, key: Iterable) -> Expression:
```

Comment:
Might want to ponder that.

```suggestion
    def __getitem__(self, key: Iterable[Union[str, Expression]]) -> Expression:
```